Custom rate and named standard rate for units (Issue #1216)

gyordong opened the pull request:
# Description

(Please include a summary of the change and which issue is touched on. Please also include relevant motivation and context.)

Fixes #1216

Done in collaboration with @huss, @aduques, @AchsahJojo, @emilysarne.

## Type of change

- [ ] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

- [X] I have followed the [OED pull request](https://openenergydashboard.org/developer/pr/) ideas
- [X] I have removed text in ( ) from the issue request
- [X] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.org/developer/cla/) and each author is listed in the Description section.

## Limitations

gyordong committed:
for review

gyordong committed:
Merge remote-tracking branch 'refs/remotes/origin/development' into development

gyordong committed:
Merge branch 'OpenEnergyDashboard:development' into development

gyordong committed:
added normal unit  interval functionality

gyordong committed:
finishing remote merge

gyordong committed:
updated

gyordong committed:
added a label to custom input and made it save the last value input

gyordong committed:
idk

gyordong committed:
Merge branch 'OpenEnergyDashboard:development' into development

gyordong committed:
updated default rate

gyordong committed:
update

gyordong committed:
updated save state for custom rate to not require enter and added validation for integers in custom value

gyordong committed:
fixed integer validation along with some labels and set default value to hour on dropdown

gyordong committed:
made custom rate update to the last selected standard value, added translation for unit.sec.in.rate.enter

gyordong committed:
all functionality except a new error for non integer input and default value to custom input if nonstandard value

huss committed:
Merge remote-tracking branch 'upstream/development' into development

huss committed:
update unit modals

- Update unit edit logic for useEffect more
- Change custom sec in rate to require enter
- Disable save if not changed and incomplete custom input
- use setState and not directy set state
- consistently use === & !==
- Added SHL comments for original developers to look at and then remove
- Formatting
- Minor edit

Note create unit does not have new logic and bar may need
to be update to be similar

huss committed:
partial update to modify

Note that @gyordong worked with me on these changes.

- Also has comments with ?? on items to do.
- Not carefully tested/completed.

gyordong committed:
added handleenter() so function relies on enter to save, added comments on customratevalid(), included customratevalid in editunit's canSave useEffect check to fix a bug with negative numbers.

aduques committed:
rearranged the handleShow() and handleClose() to be in the end state

aduques committed:
Merge branch 'development' of https://github.com/gyordong/OEDdev into development

gyordong committed:
removed comments as requested by maintainer for PR

gyordong committed:
formatting, comment clean-up and added a mozilla header

gyordong committed:
Revert "Merge remote-tracking branch 'origin/1188sort' into 1216issue"

This reverts commit 1593d6d50cf36aa1272c95ec23df4fef107f1df5, reversing
changes made to 47dfa72393a96380b0fe0a5522f3705e6418b977.

gyordong committed:
Update data.ts

gyordong committed:
typo

huss commented on src/client/app/translations/data.ts:
```diff
@@ -473,6 +476,7 @@ const LocaleTranslationData = {
 		"unit.preferred.display": "Preferred Display:",
 		"unit.represent": "Unit Represent:",
 		"unit.sec.in.rate": "Sec in Rate:",
+		"unit.sec.in.rate.enter": "Input seconds in rate and then hit enter:\u{26A1}",
```
This is the en (English) section so the text is already in the correct language. Thus, no ``\u{26A1}`` is needed and should be removed.

gyordong replied:
done as requested

gyordong replied:
done as requested.

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -40,68 +42,117 @@ export default function CreateUnitModalComponent() {
 		id: -99
 	};
-
 	/* State */
```
I'm just going to note that you removed blank lines added by others. I think the intend was to leave space between sections/items of code. It was probably best to leave most of these, esp. given you left some blank lines. Up to you if you want to put them back as it will be accepted either way. FYI.

gyordong replied:
readded a line as requested

gyordong committed:
Merge branch 'OpenEnergyDashboard:development' into 1188sort

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -40,68 +42,117 @@ export default function CreateUnitModalComponent() {
+			setShowCustomInput(true);
+		} else {
+			setRate(value);
+			setState({ ...state, [e.target.name]: Number(value) });
```
This seems specific to secInRate (only used there) so should it be general or specific to that for the state update? That is what seems to happen in the handleCustomRateChange.

gyordong replied:
set to only affect secInRate instead of a general target as requested

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -40,68 +42,117 @@ export default function CreateUnitModalComponent() {
-	// Submit
-	const handleSubmit = () => {
+	// Save
+	const handleSave = () => {
```
While this new name is okay it differs from the one in edit unit of ``handleSaveChanges``. It would be nice if they were consistent.

gyordong replied:
refactored to handleSaveChanges() as requested.

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -40,68 +42,117 @@ export default function CreateUnitModalComponent() {
-		if (state.typeOfUnit != UnitType.suffix && state.suffix != '') {
-			state.typeOfUnit = UnitType.suffix;
-		}
+		const submitState = {
```
While the change in how state is set here is nice, edit unit differs. This change should probably go there too. Note identifier is set here and done after save in edit so make consistent. I think I did this so it probably isn't your doing. If you prefer, I can make the change.

aduques replied:
changed setting state in EditUnitModalComponent to be consistent with CreateUnitModalComponent

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -111,180 +162,274 @@ export default function CreateUnitModalComponent() {
+							{/* Identifier input */}
+							<Col>
+								<FormGroup>
+									<Label for="identifier">{translate('identifier')}</Label>
```
While OED is far from consistent in using single quotes over double quotes, the desire was for single quote when possible. I'm noting for the record and you don't have to change them back. At some point OED should systematically do this.

gyordong replied:
set back to single quotes as requested

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -59,8 +65,63 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
+	 * @param rate The rate to check
+	 * @returns true if the rate is custom and false if it is a standard value.
+	 */
+	const isCustomRate = (rate: number) => {
```
Crreate unit does not check for custom this way and it uses different names for some of the functions that follow. I'm wondering why create is not the same as edit. Since I made many of these changes I probably think the names are better (it may be my bias) but I do think they are more appropriate given the specific function. Do you have thoughts on this and which of the two methods might be best shared?

emilysarne replied:
Should we consider aligning the implementation of isCustom in the Create file with isCustomRate in Edit for consistency? This approach is also more efficient, as isCustomRate uses predefined rates with Object.entries. If not, could you clarify the distinction in their expected behavior to ensure we're addressing both cases as intended?

huss replied:
I agree. I was pointing out the difference and, not very clearly, indicating my preference toward doing it the way edit unit did it. I think they should use similar techniques. I was just indicating I was open to which ones if you felt there was a reason to do it another way from edit.

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -40,68 +42,117 @@ export default function CreateUnitModalComponent() {
 	/* State */
 	// Unlike EditUnitModalComponent, there are no props so we don't pass show and close via props.
 	// Modal show
+	const CUSTOM_INPUT = '-99';
```
Edit unit has near top of function. I think that is a better place for this type of const.

Also, I'm noticing now that -99 is used for the id and default graphic unit when no value. Given this, I think edit and create should use some other value for custom input. Maybe -77 (-88 looks too much like -99)? I'm hoping this is easy to do as it is a const.

gyordong replied:
changed as requested to -77.

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -132,17 +193,33 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
+		// - The rate is set so not the custom input value. This happens if select custom value but don't input with enter.
+		// - The custom rate is a positive integer
+		const validUnit = state.name !== '' &&
+			(state.typeOfUnit !== UnitType.suffix || state.suffix !== '') && state.secInRate !== Number(CUSTOM_INPUT)
```
This checks for not hitting enter on custom but create does not. This is kind to the user and nice to have.

aduques replied:
After looking at comments on the CreateUnitModalComponent, I noticed that the design of the current implementation was to discard all values when leaving the page without saving. Given this, the custom rate would be discarded alongside the other values despite changing the similar line of code in CreateModalComponent. Should CreateUnitModalComponent be changed to still keep unsaved changes similarly to how it is implemented in EditUnitModalComponent?

huss replied:
I thought this test was to see if the user had started a custom input but not hit enter. This likely indicates they started to enter a value but never finalized it with an enter. This seems a potentially common mistake that OED is trying to catch to avoid. I didn't think it has to do with leaving the page and resetting state but maybe I'm missing something. What do you think?

gyordong replied:
Hi Steve, I changed the code to check for enter. The only caveat is that I made it purposely allow users to save without clicking enter if they didn't change the custom value from the standard values automatically inputted when clicked (ex. day to custom value, box shows 86400, create unit can be clicked). Let me know if this doesn't seem ideal and I'll change it to require enter to create a unit under all circumstances.

huss replied:
Sorry I missed this comment until now. I'm not sure which is best. It is unclear why the admin selected custom if they did not intend to change the value. However, I'm okay with this as long as both create and edit work the same.

gyordong replied:
I think the difference won't be noticeable for the user. Should I change it to require a change of value or is it fine as is?

huss replied:
I'm not sure if I understand but it is okay to use the value without enter if it is the standard value as long as both create and edit do the same thing. Is this what you were asking?

gyordong replied:
Yes. Do you have a preference between the two implementations?

huss replied:
I don't care too much but I might pick that if they do custom then they have to hit enter. One case I am thinking about is if they change the value away from a standard one and then put it back without hitting enter then would the code accept this second value because it is standard? This type of issue is why I'm leaning toward requiring enter for all custom values but maybe it isn't an issue.

gyordong replied:
Okay, I've edited it so it requires both that the custom value isn't a standard rate and that enter is pressed. I believe this makes it line up with edit.

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -40,68 +42,117 @@ export default function CreateUnitModalComponent() {
+		resetCustomRate();
 	};
 
+	const handleShow = () => setShowModal(true);
```
Not your doing but could you add braces around the one function line of code? OED prefers them in all cases (see edit).

gyordong replied:
added braces as requested

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -111,180 +162,274 @@ export default function CreateUnitModalComponent() {
+										id="suffix"
+										name="suffix"
+										type="text"
+										autoComplete="off"
```
This isn't in edit. Should it be?

gyordong replied:
Not sure, since suffix is a unit type I think autocomplete might serve as quality of life for users who don't want to have to retype for meters that share the same unit. Then again, I'm not familiar with the purpose of suffix. I would love some additional info on how it works so I can give more informed feedback.

huss replied:
Suffix is a string that is the end of a special type of unit's name. It seems unlikely that it would be reused or match a standard used string. I think it is probably best not to allow autocomplete in this case.

gyordong replied:
removed as requested

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -279,157 +350,201 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
+						</Row>
+						{/* Note input */}
+						<FormGroup>
+							<Label for='note'>{translate('unit')}</Label>
```
The translate should be to note (as in create).

gyordong replied:
changed as requested

huss commented on src/client/app/components/unit/EditUnitModalComponent.tsx:
```diff
@@ -279,157 +350,201 @@ export default function EditUnitModalComponent(props: EditUnitModalComponentProp
+								name='note'
+								type='textarea'
+								value={state.note}
+								placeholder='Note'
```
There are several placeholder values in edit but not create. I think I prefer them to be gone so could you do that?

gyordong replied:
removed as requested

gyordong replied:
I know you asked me to remove placeholders, but removing the placeholder='Identifier' in editunitmodalcomponent.tsx caused a testcase to break.
Should I just leave it in or keep the placeholder until someone fixes the testcase as TODO?
the test in question was: 
1) Test CSV Pipeline
         Testing files starting 'pipe3' doing 'Cumulative time readings' with 1 requests:

huss replied:
@gyordong I thought this was strange and I looked at the code. The test code should be making a request to the server and not be impacted by the client code. Also, removing the placeholder text, I would think, would not impact the actual values on the client. I looked at this PR code on my machine and I see the commit where you removed the placeholder in EditUnitModalComponent. When I run the tests in src/server/test/web/csvPipelineTest.js they all pass. I even ran all the tests and they all pass. Thus, I'm not sure I am seeing what you see. Do you know what is up since I'm now confused.

gyordong replied:
It passes now as well, even if with placeholder removed. May have just been a glitch on my local machine.

huss reviewed (changes requested):
I want to thank @gyordong, @aduques, @AchsahJojo & Emily Sarne for this contribution. Is Emily Sarne the GitHub user @emilysarne? I want to verify they have signed the CLA if they contributed to this work.

Testing found this works as desired. Most of my comments relate to differences between the code for edit and create for unit. I welcome your thoughts on my comments and you can address if you think a comment is fine. Please let me know if I can help in any way.

gyordong committed:
syntax changes and clean up

gyordong committed:
more syntax changes

gyordong committed:
refactored handlesave

gyordong committed:
made function specific to secinrate and updated customrate value

gyordong committed:
@emilysarne implemented isCustomRate() in editunitmodalcomponent.tsx, I removed suffix

aduques committed:
Update EditUnitModalComponent.tsx

changed setting state in EditUnitModalComponent to be consistent with CreateUnitModalComponent

aduques committed:
fixed syntax

aduques committed:
Merge branch '1188sort' of https://github.com/gyordong/OEDdev into 1188sort

aduques committed:
removed line that was added from earlier merge

gyordong committed:
added check for enter on createunitmodalcomponent.tsx

gyordong committed:
added enter on createunit pt2

gyordong committed:
create checks for nonstandard rate

aduques committed:
Merge branch 'OpenEnergyDashboard:development' into 1188sort

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -45,63 +48,129 @@ export default function CreateUnitModalComponent() {
-		setState({ ...state, [e.target.name]: Number(e.target.value) });
+	const [canSave, setCanSave] = useState(false);
+	// Sets the starting rate for secInRate box, value of 3600 is chosen as default to result in Hour as default in dropdown box.
+	const isCustomRate = (rate: number) => {
```
The comment on the line above belongs with the line after this function. Also, please copy the comment in edit about this function to be the comment for it.

huss replied:
I've done this change.

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -45,63 +48,129 @@ export default function CreateUnitModalComponent() {
+  */
+	const handleStandardNumberChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+		const { value } = e.target;
+		const isCustom = isCustomRate(Number(value));
```
isCustom is used in both parts of the if statement below in the same way. Given this, I would move this line right after this line and before the if (and remove from the two parts of the if statement).

huss replied:
This is gone given my other changes.

huss commented on src/client/app/components/unit/CreateUnitModalComponent.tsx:
```diff
@@ -111,180 +180,273 @@ export default function CreateUnitModalComponent() {
+										id="secInRate"
+										name="secInRate"
+										type="select"
+										onChange={e => handleStandardNumberChange(e)}
```
To be consistent with edit, I think this should be handleRateChange.

huss replied:
Done with my edits.

huss committed:
align edit/create unit + other

- Keep state declaration at top of unit modals
- Place isCustomRate in utils to share
- reorder some lines of code so consistent
- Remove comment on edit reset in other files
- Other minor edits

huss committed:
Merge remote-tracking branch 'gyordong/1188sort' into pr/gyordong/1388

huss reviewed (approved):
@gyordong The changes you made looked good and worked in almost every case. Since it was complex to note all my ideas and to move this forward, I pushed a commit that aligned unit edit and create. With this, I think this PR is now ready to merge. I'll let it sit for a little while to give everyone involved a chance to look at what I did and comment if they have nay concerns. If nothing needs changing then it will be merged as is. Congratulations on completing this work and I appreciate your efforts to do address all the comments given changes over time.
